src/Storages/MergeTree/MergeTask.cpp

```diff
@@ -877,6 +879,8 @@ void MergeTask::ExecuteAndFinalizeHorizontalPart::createMergedStream()
     global_ctx->horizontal_stage_progress = std::make_unique<MergeStageProgress>(
         ctx->column_sizes ? ctx->column_sizes->keyColumnsWeight() : 1.0);
 
+    global_ctx->merging_column_names.emplace_back(BlockNumberColumn.name);
```

Comment:
Merging columns are PK columns: https://github.com/ClickHouse/ClickHouse/blob/a50e741bea349775154278fb39a5afd290174ee8/src/Storages/MergeTree/MergeTask.cpp#L84-L96
gathering columns are non sorting column. So for horizontal merge BlockNumberColumn should be merging_column and for vertical merge it should be gathering_columns: https://github.com/ClickHouse/ClickHouse/blob/a50e741bea349775154278fb39a5afd290174ee8/src/Storages/MergeTree/MergeTask.cpp#L278-L303